lib/core/src/Cardano/Wallet/Primitive/Scripts.hs

```diff
@@ -0,0 +1,105 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- An implementation of wallet's verification key discovery
+-- in a script.
+
+module Cardano.Wallet.Primitive.Scripts
+    ( isShared
+    , retrieveAllVerKeyHashes
+    ) where
+
+import Prelude
+
+import Cardano.Address.Script
+    ( KeyHash (..), Script (..), ScriptHash (..), toScriptHash )
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..), Index (..), Role (..), SoftDerivation (..) )
+import Cardano.Wallet.Primitive.AddressDerivation.Shelley
+    ( ShelleyKey (..) )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( SeqState (..)
+    , mkVerificationKeyPool
+    , verPoolAccountPubKey
+    , verPoolGap
+    , verPoolIndexedKeys
+    , verPoolKnownScripts
+    )
+import Cardano.Wallet.Primitive.Types.Address
+    ( AddressState (..) )
+import Control.Monad
+    ( foldM )
+import Data.Coerce
+    ( coerce )
+import Data.Functor.Identity
+    ( Identity (..) )
+
+import qualified Data.List as L
+import qualified Data.Map.Strict as Map
+
+instance Ord ScriptHash where
+    compare (ScriptHash sh1) (ScriptHash sh2) = compare sh1 sh2
+
+isShared
+    :: (k ~ ShelleyKey, SoftDerivation k)
+    => Script
+    -> SeqState n k
+    -> ([k 'ScriptK XPub], SeqState n k)
+isShared script (SeqState !s1 !s2 !ixs !rpk !prefix !s3) =
+    let verKeysInScript = retrieveAllVerKeyHashes script
+        toVerKey = deriveAddressPublicKey (verPoolAccountPubKey s3) MultisigScript
+        ourVerKeysInScript =
+            map (\(_,(ix,_)) -> toVerKey (coerce ix) ) $
+            filter (\(keyH,_) -> keyH `elem` verKeysInScript) $
+            Map.toList (verPoolIndexedKeys s3)
+        scriptXPubs = L.nub $ map coerce ourVerKeysInScript
+        updateAddressState k current =
+            if k `elem` verKeysInScript then
+                Used
+            else
+                current
+        indexedKeys =
+            Map.mapWithKey (\keyH (ix,isUsed) -> (ix, updateAddressState keyH isUsed) )
+            (verPoolIndexedKeys s3)
+        insertIf predicate k v = if predicate v then Map.insert k v else id
+        knownScripts =
+            insertIf (not . null) (toScriptHash script) scriptXPubs
+            (verPoolKnownScripts s3)
+
+        -- if there are no gap number of consecutive NotUsed verification keys
+        -- then we extend the verification key pool
+        s3' = mkVerificationKeyPool (verPoolAccountPubKey s3) (verPoolGap s3)
+              indexedKeys knownScripts
+    in ( scriptXPubs
+       , SeqState s1 s2 ixs rpk prefix s3'
+       )
```

Comment:
I'll suggest defining two intermediate functions such as: 

```hs
lookupKeyHash
    :: KeyHash
    -> VerificationKeyPool k
    -> (Maybe (Index 'Soft 'ScriptK), VerificationKeyPool k)

modifyKnownScripts
    :: (Map ScriptHash [Index 'Soft 'ScriptK] -> Map ScriptHash [Index 'Soft 'ScriptK])
    -> VerificationKeyPool k
    -> VerificationKeyPool k
```

Which allows rewriting this function in the following form:

```hs
isShared
    :: (k ~ ShelleyKey, SoftDerivation k)
    => Script
    -> SeqState n k
    -> ([k 'ScriptK XPub], SeqState n k)
isShared script (SeqState !s1 !s2 !pending !rpk !prefix !scriptPool) =
    let
        hashes = foldScript (:) [] script

        (ixs, scriptPool') = scriptPool
            & runState (mapM (state . lookupKeyHash) hashes)
            & first catMaybes
            & \(ixs, scriptPool') ->
                ( ixs
                , modifyKnownScripts
                    (insertIf (not . null) (toScriptHash script) ixs)
                    scriptPool'
                )
    in
        ( deriveAddressPublicKey (verPoolAccountPubKey scriptPool') MultisigScript <$> ixs
        , SeqState s1 s2 pending rpk prefix scriptPool'
        )
```

The rationale is that `lookupKeyHash` would be  a function defined in `Cardano.Wallet.Primitive.AddressDiscovery.Sequential` which leverages the internal maps in `VerificationKeyPool` to lookup a single key and extend the map appropriately, very much like `lookupAddress` does for `AddressPool`. It is testable independently and self contained (it doesn't also leak the internals of a 'VerificationKeyPool' away and take care of ensuring the envariant regarding the pool gap). 

Such function would have the following signature:
Also, on a more minor note, I'd suggest to move `insertIf` out of the local declarations to enhance readability:

```hs
insertIf :: Ord k => (v -> Bool) -> k -> v -> Map k v -> Map k v
insertIf predicate k v =
    if predicate v then Map.insert k v else id
```

Also, note that you'll need to tweak a bit the `SoftDerivation` class declaration to make the `role` parameter flexible and avoid all those coercion altogether:

```hs
class HardDerivation key => SoftDerivation (key :: Depth -> * -> *) where
    -- | Derives address public key from the given account public key, using
    -- derivation scheme 2 (see <https://github.com/input-output-hk/cardano-crypto/ cardano-crypto>
    -- package for more details).
    --
    -- This is the preferred way of deriving new sequential address public keys.
    deriveAddressPublicKey
        :: key 'AccountK XPub
        -> Role
        -> Index 'Soft rol
        -> key rol XPub
```

Reply:
lookupKey introduced here https://github.com/input-output-hk/cardano-wallet/pull/2308/commits/de022a5c0af14a4410ecfd256df0e1151683c744